Index: loginpassword.cpp
===================================================================
--- loginpassword.cpp	(Revision 13632)
+++ loginpassword.cpp	(Arbeitskopie)
@@ -58,6 +58,7 @@
 			SetDlgItemTextUtf(hwndDlg, IDC_INSTRUCTION, pszUIN);
 
 			SendDlgItemMessage(hwndDlg, IDC_LOGINPW, EM_LIMITTEXT, PASSWORDMAXLEN - 1, 0);
+			SendMessage(hwndDlg, WM_NEXTDLGCTL, IDC_LOGINPW, true);
 
 			CheckDlgButton(hwndDlg, IDC_SAVEPASS, ppro->getSettingByte(NULL, "RememberPass", 0));
 		}
