Index: jabber_opt.cpp
===================================================================
--- jabber_opt.cpp	(Revision 13589)
+++ jabber_opt.cpp	(Arbeitskopie)
@@ -743,12 +743,13 @@
 	{
 		m_gotservers = node ? true : false;
 
-		TCHAR *server = m_cbServer.GetText();
-		bool bDropdown = m_cbServer.GetDroppedState();
-		if (bDropdown) m_cbServer.ShowDropdown(false);
+		if ( node ) {
+			TCHAR *server = m_cbServer.GetText();
+			bool bDropdown = m_cbServer.GetDroppedState();
+			if (bDropdown) m_cbServer.ShowDropdown(false);
 
-		m_cbServer.ResetContent();
-		if ( node ) {
+			m_cbServer.ResetContent();
+			
 			for (int i = 0; ; ++i) {
 				HXML n = xmlGetChild(node, i);
 				if ( !n )
@@ -759,12 +760,12 @@
 						if ( m_cbServer.FindString(jid, -1, true) == CB_ERR)
 							 m_cbServer.AddString(jid);
 			}
+			
+			m_cbServer.SetText(server);
+
+			if (bDropdown) m_cbServer.ShowDropdown();
+			mir_free(server);
 		}
-
-		m_cbServer.SetText(server);
-
-		if (bDropdown) m_cbServer.ShowDropdown();
-		mir_free(server);
 	}
 
 	static void QueryServerListThread(void *arg)
@@ -2202,13 +2203,14 @@
 {
 	m_gotservers = node ? true : false;
 
-	TCHAR *server = m_cbServer.GetText();
-	bool bDropdown = m_cbServer.GetDroppedState();
-	if (bDropdown) m_cbServer.ShowDropdown(false);
-
-	m_cbServer.ResetContent();
 	if ( node )
 	{
+		TCHAR *server = m_cbServer.GetText();
+		bool bDropdown = m_cbServer.GetDroppedState();
+		if (bDropdown) m_cbServer.ShowDropdown(false);
+
+		m_cbServer.ResetContent();
+		
 		for (int i = 0; ; ++i) {
 			HXML n = xmlGetChild(node, i);
 			if ( !n )
@@ -2219,12 +2221,12 @@
 					if (m_cbServer.FindString(jid, -1, true) == CB_ERR)
 						m_cbServer.AddString(jid);
 		}
+		
+		m_cbServer.SetText(server);
+		
+		if (bDropdown) m_cbServer.ShowDropdown();
+		mir_free(server);
 	}
-
-	m_cbServer.SetText(server);
-
-	if (bDropdown) m_cbServer.ShowDropdown();
-	mir_free(server);
 }
 
 void CJabberDlgAccMgrUI::QueryServerListThread(void *arg)
