Index: msgs.h
===================================================================
--- msgs.h	(revision 10988)
+++ msgs.h	(working copy)
@@ -323,6 +323,8 @@
 #define SRMSGSET_MSGTIMEOUT_MIN    4000 // minimum value (4 seconds)
 #define SRMSGSET_FLASHCOUNT        "FlashMax"
 #define SRMSGDEFSET_FLASHCOUNT     3
+#define SRMSGSET_LASTFLASH        "LastFlash"
+#define SRMSGDEFSET_LASTFLASH     0
 
 #define SRMSGSET_LOADHISTORY       "LoadHistory"
 #define SRMSGDEFSET_LOADHISTORY    LOADHISTORY_UNREAD
Index: msgwindow.c
===================================================================
--- msgwindow.c	(revision 10988)
+++ msgwindow.c	(working copy)
@@ -740,6 +740,8 @@
 			} else {// || ((GetActiveWindow() == hwndDlg) && (GetForegroundWindow() == hwndDlg))) {
 				KillTimer(hwndDlg, TIMERID_FLASHWND);
 				FlashWindow(hwndDlg, FALSE);
+				if (DBGetContactSettingByte(NULL, SRMMMOD, SRMSGSET_LASTFLASH, SRMSGDEFSET_LASTFLASH)!=0)
+					FlashWindow(hwndDlg, TRUE);	// make blink
 			}
 		}
 		break;
