Index: miranda/protocols/IcqOscarJ/icq_opts.cpp
===================================================================
--- miranda/protocols/IcqOscarJ/icq_opts.cpp	(Revision 13869)
+++ miranda/protocols/IcqOscarJ/icq_opts.cpp	(Arbeitskopie)
@@ -443,16 +443,24 @@
 		switch (LOWORD(wParam)) {
 		case IDC_UTFENABLE:
 			icq_EnableMultipleControls(hwndDlg, icqUnicodeControls, SIZEOF(icqUnicodeControls), IsDlgButtonChecked(hwndDlg, IDC_UTFENABLE));
+			OptDlgChanged(hwndDlg);
 			break;
+		case IDC_UTFCODEPAGE:
+			if(HIWORD(wParam)==CBN_SELCHANGE)
+				OptDlgChanged(hwndDlg);
+			break;
 		case IDC_DCENABLE:
 			icq_EnableMultipleControls(hwndDlg, icqDCMsgControls, SIZEOF(icqDCMsgControls), IsDlgButtonChecked(hwndDlg, IDC_DCENABLE));
+			OptDlgChanged(hwndDlg);
 			break;
 		case IDC_XSTATUSENABLE:
 			icq_EnableMultipleControls(hwndDlg, icqXStatusControls, SIZEOF(icqXStatusControls), IsDlgButtonChecked(hwndDlg, IDC_XSTATUSENABLE));
     case IDC_MOODSENABLE:
       icq_EnableMultipleControls(hwndDlg, icqCustomStatusControls, SIZEOF(icqCustomStatusControls), IsDlgButtonChecked(hwndDlg, IDC_XSTATUSENABLE) || IsDlgButtonChecked(hwndDlg, IDC_MOODSENABLE));
+		default:
+			OptDlgChanged(hwndDlg);
+			break;
 		}
-		OptDlgChanged(hwndDlg);
 		break;
 
 	case WM_NOTIFY:
Index: miranda/protocols/IcqOscarJ/icq_popups.cpp
===================================================================
--- miranda/protocols/IcqOscarJ/icq_popups.cpp	(Revision 13869)
+++ miranda/protocols/IcqOscarJ/icq_popups.cpp	(Arbeitskopie)
@@ -58,11 +58,13 @@
 
 INT_PTR CALLBACK DlgProcIcqPopupOpts(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam)
 {
+	static bool bInitDone = true;
 	BYTE bEnabled;
 	CIcqProto* ppro = (CIcqProto*)GetWindowLongPtr( hwndDlg, GWLP_USERDATA );
 
 	switch (msg) {
 	case WM_INITDIALOG:
+		bInitDone = false;
 		ICQTranslateDialog(hwndDlg);
 
 		ppro = (CIcqProto*)lParam;
@@ -93,6 +95,7 @@
 		CheckDlgButton(hwndDlg, IDC_POPUPS_ENABLED, bEnabled);
 		icq_EnableMultipleControls(hwndDlg, icqPopupsControls, SIZEOF(icqPopupsControls), bEnabled);
 		icq_EnableMultipleControls(hwndDlg, icqPopupColorControls, SIZEOF(icqPopupColorControls), bEnabled & !IsDlgButtonChecked(hwndDlg,IDC_USEWINCOLORS));
+		bInitDone = true;
 		return TRUE;
 
 	case WM_COMMAND:
@@ -114,9 +117,20 @@
 		case IDC_USEWINCOLORS:
 			bEnabled = IsDlgButtonChecked(hwndDlg,IDC_POPUPS_ENABLED);
 			icq_EnableMultipleControls(hwndDlg, icqPopupColorControls, SIZEOF(icqPopupColorControls), bEnabled & !IsDlgButtonChecked(hwndDlg,IDC_USEWINCOLORS));
+			SendMessage(GetParent(hwndDlg), PSM_CHANGED, 0, 0);
 			break;
+		case IDC_POPUP_LOG0_TIMEOUT:
+		case IDC_POPUP_LOG1_TIMEOUT:
+		case IDC_POPUP_LOG2_TIMEOUT:
+		case IDC_POPUP_LOG3_TIMEOUT:
+		case IDC_POPUP_SPAM_TIMEOUT:
+			if((HIWORD(wParam) == EN_CHANGE) && bInitDone)
+				SendMessage(GetParent(hwndDlg), PSM_CHANGED, 0, 0);
+			break;
+		default:
+			SendMessage(GetParent(hwndDlg), PSM_CHANGED, 0, 0);
+			break;
 		}
-		SendMessage(GetParent(hwndDlg), PSM_CHANGED, 0, 0);
 		break;
 
 	case WM_NOTIFY:
