Index: jabber_proto.cpp
===================================================================
--- jabber_proto.cpp	(revision 11537)
+++ jabber_proto.cpp	(working copy)
@@ -1273,6 +1273,17 @@
 						GcQuit(item, 0, NULL);
 			}
 
+			TCHAR szMessage[ 512 ];
+			time_t rawtime;
+			struct tm * timeinfo;
+			time ( &rawtime );
+			timeinfo = localtime ( &rawtime );
+			wsprintf(szMessage, _T("offline since %.2i.%.2i.%.4i %.2i:%.2i:%.2i"), timeinfo->tm_mday, timeinfo->tm_mon + 1, timeinfo->tm_year + 1900, 
+						timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec);
+			m_ThreadInfo->send(
+			XmlNode( _T("presence")) << XATTR( _T("type"), _T("unavailable")) 
+					<< XCHILD( _T("status"), szMessage));
+
 			m_ThreadInfo->send( "</stream:stream>" );
 			m_ThreadInfo->close();
 			m_ThreadInfo = NULL;
